hypothesis-python/src/hypothesis/extra/gufunc.py

```diff
@@ -0,0 +1,451 @@
+        x = int(arg)
+        assert arg == x  # e.g., check 5.0 and not 5.5 was passed
+    except Exception:
+        raise InvalidArgument("%s=%r (type=%s) not representable as int"
+                              % (name, arg, type(arg).__name__))
+    return x
+
+
+def _int_or_dict(x, default_val):
+    """Pre-process cases where argument `x` can be `int`, `dict`, or
+    `defaultdict`. In all cases, build a `defaultdict` and return it.
+    """
+    # case 1: x already defaultdict, leave it be, pass thru
+    if isinstance(x, defaultdict):
+        return x
+
+    default_val = ensure_int(default_val, "default value")
+    try:
+        # case 2: x is or can be converted to dict
+        D = defaultdict(lambda: default_val, x)
+    except TypeError:
+        # case 3: x is or can be converted to int => make a const dict
+        default_val = ensure_int(x, "constant value")
+        D = defaultdict(lambda: default_val)
+    # case 4: if can't be converted to dict or int, then exception raised
+    return D
+
+
+@composite
+def _arrays(draw, dtype, shape, elements=None, unique=False):
+    """Wrapper to fix issues with `hypothesis.extra.numpy.arrays`.
+
+    `arrays` is strict on shape being `int` which this fixes. This is partially
+    not needed in Py3 since there is no `int` vs `long` issue. Also, `arrays`
+    does not return ndarray for 0-dim arrays.
+
+    This could possibly be done more compactly with:
+    ```
+    S = builds(np.asarray,
+               arrays(dtype, shape, elements=elements, unique=unique),
+               dtype=just(dtype))
+    ```
+    However, this appears to be slower.
+    """
+    # Maintain ability of arrays to accept strategies for these args
+    if isinstance(dtype, SearchStrategy):
+        dtype = draw(dtype)
+    if isinstance(shape, SearchStrategy):
+        shape = draw(shape)
+
+    shape = tuple(ensure_int(aa) for aa in shape)
+    S = arrays(dtype, shape, elements=elements, unique=unique).map(np.asarray)
+    X = draw(S)
+    X = X.astype(dtype, copy=False)  # Will never see original => copy=False
+    return X
+
+
+@composite
+def _tuple_of_arrays(draw, shapes, dtype, elements, unique=False):
+    """Strategy to generate a tuple of ndarrays with specified shapes.
+
+    Parameters
+    ----------
+    shapes : list of tuples of int
+        List of tuples where each tuple is the shape of an argument. A
+        `SearchStrategy` for list of tuples is also supported.
+    dtype : list-like of dtype
+        List of numpy `dtype` for each argument. These can be either strings
+        (``'int64'``), type (``np.int64``), or numpy `dtype`
+        (``np.dtype('int64')``). Built in Python types (`int`, `float`, etc)
+        also work. A single `dtype` can be supplied for all arguments.
+    elements : list-like of strategy
+        Strategies to fill in array elements on a per argument basis. One can
+        also specify a single strategy
+        (e.g., :func:`hypothesis.strategies.floats`)
```

Comment:
:func:`~hypothesis.strategies.floats`

(adding the `~` will render this cross-reference as `floats()` rather than the full qualname)

Reply:
ok